src/coreclr/System.Private.CoreLib/src/System/Runtime/ControlledExecution.CoreCLR.cs

```diff
@@ -0,0 +1,173 @@
+        /// <exception cref="System.ArgumentNullException">The <paramref name="action"/> argument is null.</exception>
+        /// <exception cref="System.InvalidOperationException">
+        /// The current thread is already running the <see cref="ControlledExecution.Run"/> method.
+        /// </exception>
+        /// <exception cref="System.OperationCanceledException">The execution was aborted.</exception>
+        [Obsolete(Obsoletions.ControlledExecutionRunMessage, DiagnosticId = Obsoletions.ControlledExecutionRunDiagId, UrlFormat = Obsoletions.SharedUrlFormat)]
+        public static void Run(Action action, CancellationToken cancellationToken)
+        {
+            ArgumentNullException.ThrowIfNull(action);
+            var execution = new Execution(action, cancellationToken);
+            cancellationToken.Register(execution.Abort, useSynchronizationContext: false);
+            execution.Run();
+        }
+
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "ThreadNative_Abort")]
+        private static partial void AbortThread(ThreadHandle thread);
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static extern bool ResetAbortThread();
+
+        private sealed partial class Execution
+        {
+            // The state transition diagram (S means the Started flag and so on):
+            // N ⟶  S  ⟶ SF
+            // ↓     ↓
+            // AR   SAR ⟶ SFAR
+            // ↓     ↓       ↓
+            // A    SA  ⟶ SFA
+            private enum State : int
+            {
+                None = 0,
+                Started = 1,
+                Finished = 2,
+                AbortRequested = 4,
+                RunningAbort = 8
+            }
+
+            [ThreadStatic]
+            private static Execution? t_execution;
+
+            // Interpreted as a value of the State enumeration type
+            private int _state;
+            private readonly Action _action;
+            private readonly CancellationToken _cancellationToken;
+            private Thread? _thread;
+
+            public Execution(Action action, CancellationToken cancellationToken)
+            {
+                _action = action;
+                _cancellationToken = cancellationToken;
+            }
+
+            public void Run()
+            {
+                Debug.Assert((_state & (int)State.Started) == 0 && _thread == null);
+
+                // Recursive ControlledExecution.Run calls are not supported
+                if (t_execution != null)
+                    throw new InvalidOperationException(SR.InvalidOperation_NestedControlledExecutionRun);
+
+                _thread = Thread.CurrentThread;
+
+                try
+                {
+                    try
+                    {
+                        // As soon as the Started flag is set, this thread may be aborted asynchronously
+                        if (Interlocked.CompareExchange(ref _state, (int)State.Started, (int)State.None) == (int)State.None)
```

Comment:
The `Abort` method may be called before the `_state` is set in the `Run` method.

Reply:
Yeah, I realized that subsequently... but that's only because of when Register is being called.
https://github.com/dotnet/runtime/pull/71661#discussion_r927930923